### PR TITLE
docs(adrs): add ADR-055 content-based violation fingerprint suppression

### PR DESCRIPTION
**Pull request by cidrblock**
## Summary

- Propose ADR-055: content-based SHA-256 fingerprinting for violations, enabling permanent accept/ignore decisions that survive line-number changes, reformatting, and transfer across projects
- Default fingerprint: `hash(rule_id + normalized_full_task_yaml)` — the entire task content is the identity, not just the name or module
- Fingerprinting is a consumer-side concern (Gateway/CLI), not an engine change — the engine already emits `rule_id` and `original_yaml` on every violation
- Establishes rule ID permanence policy: retired IDs are reserved, never reused

## Changes

- `.sdlc/adrs/ADR-055-violation-fingerprint-suppression.md` — full ADR with context, decision, alternatives considered, consequences, and phased implementation notes
- `.sdlc/adrs/README.md` — index updated with new entry in Proposed section

## Key design decisions

- **Full task YAML is the hash input** — different variable values (e.g. `apt install nginx` vs `rm -rf /data`) must produce different fingerprints because accepting one should not accept the other
- **Normalize whitespace and comments before hashing** — these are not runtime values and should not invalidate acceptances when `apme format` runs
- **Engine is not involved** — preserves invariant 5 (stateless engine) and invariant 11 (engine only emits)
- **User-tunable granularity** — advanced option to loosen the fingerprint (rule+module, rule-only), but the default is maximally specific
- **Dual storage** — Gateway DB for enterprise, `.apme/suppressions.yml` CLI file for standalone/CI

## Test plan

- [ ] `tox -e lint` passes
- [ ] ADR follows template structure and references existing ADRs correctly
- [ ] No code changes — documentation only

### CODE REVIEW

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+Where:
+- `rule_id` is the violation's rule identifier (e.g. `L046`)
+- `original_yaml` is the full node YAML as originally written (already carried on the `Violation` proto, field 16)
+- `normalize()` strips non-runtime content before hashing:
+  - Strip leading/trailing whitespace per line
+  - Strip YAML comment-only lines
+  - Collapse runs of blank lines
+  - Do NOT re-order keys, change quoting style, or alter values
+
+Whitespace and comments have zero effect on what Ansible executes. Stripping them means `apme format`, re-indentation, and comment edits do not invalidate acceptances. Any change to runtime content (variable values, module arguments, conditions, task name) changes the fingerprint and forces re-review.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:70]: Good catch. Rewrote the `normalize()` spec to use a YAML-structure-aware normalizer (e.g. `ruamel.yaml` round-trip) that preserves block scalar (`|`/`>`) bodies verbatim. Lines inside block scalars that look like comments are now explicitly recognized as runtime content. Also added block scalar edge cases to the Phase 1 unit test list. Fixed in d09ca78.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+### 1. Fingerprint formula
+
+```
+SHA-256(rule_id + "\x00" + normalize(original_yaml))
+```
+
+Where:
+- `rule_id` is the violation's rule identifier (e.g. `L046`)
+- `original_yaml` is the full node YAML as originally written (already carried on the `Violation` proto, field 16)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:60]: Agreed. Added an explicit canonicalization step to the `rule_id` definition: legacy namespace prefixes (e.g. `native:L046` → `L046`) are stripped before hashing. Also added a `canonicalize_rule_id()` helper to the Phase 1 implementation notes. Fixed in d09ca78.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+## Decision
+
+**We will compute a content-based SHA-256 fingerprint for each violation from the rule ID and the full normalized task YAML, with suppression evaluated at the consumer layer (Gateway or CLI), not in the engine.**
+
+### 1. Fingerprint formula
+
+```
+SHA-256(rule_id + "\x00" + normalize(original_yaml))
+```
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:56]: Good point on the inconsistency. Updated the granularity mode table to explicitly show the full `SHA-256(rule_id + "\x00" + ...)` formula for all three modes (`full`, `rule_module`, `rule_only`), with an introductory note explaining the null byte separator prevents ambiguous concatenation. Fixed in d09ca78.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+| `fingerprint_hash` | text | SHA-256 hex digest |
+| `fingerprint_mode` | text | `"full"`, `"rule_module"`, or `"rule_only"` |
+| `rule_id` | text | Denormalized for fast lookup and human readability |
+| `scope` | text | `"global"` or `"project:<uuid>"` |
+| `reason` | text | Human-provided justification |
+| `created_by` | text | User or system that created the suppression |
+| `created_at` | text | ISO 8601 timestamp |
+
+Query: given violation fingerprints from a scan, `WHERE fingerprint_hash IN (...)` — simple indexed lookup.
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md]: Good catch. Updated the suppression lookup query to include scope filtering: `WHERE fingerprint_hash IN (...) AND (scope = 'global' OR scope = 'project:<uuid>')`. Also added a recommendation for a composite index on `(fingerprint_hash, scope)` to keep lookups efficient. Fixed in 23deb18.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+All modes use the `\x00` null byte as a field separator to prevent collisions between rule IDs and content that could otherwise concatenate ambiguously:
+
+| Mode | Hash input | Use case |
+|------|-----------|----------|
+| `full` (default) | `SHA-256(rule_id + "\x00" + normalize(original_yaml))` | "I accept this exact content for this rule" |
+| `rule_module` | `SHA-256(rule_id + "\x00" + module_fqcn)` | "I accept all tasks using this module for this rule" |
+| `rule_only` | `SHA-256(rule_id + "\x00")` | "I never care about this rule" (overlaps ADR-041 disable, useful in CLI-file workflows) |
+
+The suppression record stores which mode was used, so the matching logic knows how to compare. The default is maximally specific (safest); users explicitly broaden scope when they understand the implications.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:97]: Good catch. Added explicit `mode` field to the CLI suppressions.yml example. When omitted it defaults to `full` for backward compatibility. Fixed in f9bd5c0.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+## Decision
+
+**We will compute a content-based SHA-256 fingerprint for each violation from the rule ID and the full normalized task YAML, with suppression evaluated at the consumer layer (Gateway or CLI), not in the engine.**
+
+### 1. Fingerprint formula
+
+```
+SHA-256(rule_id + "\x00" + normalize(original_yaml))
+```
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:57]: Agreed — cross-platform consistency requires an explicit encoding contract. Added a "Canonical encoding for hash input" paragraph specifying UTF-8 encoding, Unicode NFC normalization, and LF line endings (CRLF converted to LF before hashing). Fixed in f9bd5c0.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+suppressions:
+  - fingerprint: "a1b2c3d4..."
+    rule_id: L046
+    reason: "Accepted: legacy module usage in bootstrap role"
+    created: "2026-05-21"
+  - fingerprint: "e5f6a7b8..."
+    rule_id: M014
+    reason: "Deferred: migrating ansible_hostname in Phase 3"
+    created: "2026-05-21"
+```
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md:147]: Good point. Added a top-level `version: 1` field to the CLI file format example. This enables deterministic migration if the fingerprint algorithm or normalization ever needs to change. Fixed in f9bd5c0.

File: .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+Power users can opt into looser fingerprints by choosing which components to include. This is an advanced option, not the default:
+
+All modes use the `\x00` null byte as a field separator to prevent collisions between rule IDs and content that could otherwise concatenate ambiguously:
+
+| Mode | Hash input | Use case |
+|------|-----------|----------|
+| `full` (default) | `SHA-256(rule_id + "\x00" + normalize(original_yaml))` | "I accept this exact content for this rule" |
+| `rule_module` | `SHA-256(rule_id + "\x00" + module_fqcn)` | "I accept all tasks using this module for this rule" |
+| `rule_only` | `SHA-256(rule_id + "\x00")` | "I never care about this rule" (overlaps ADR-041 disable, useful in CLI-file workflows) |
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply .sdlc/adrs/ADR-055-violation-fingerprint-suppression.md]: Added an explicit resolution order: prefer `metadata.resolved_fqcn` (the post-redirect fully-qualified name), fall back to `metadata.original_module` when resolution is unavailable. This keeps fingerprints stable across collection redirect changes. Fixed in f9bd5c0.
